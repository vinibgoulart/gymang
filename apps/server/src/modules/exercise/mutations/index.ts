import ExerciseAdd from './ExerciseAddMutation';

const ExerciseMutations = {
  ExerciseAdd,
};

export default ExerciseMutations;
