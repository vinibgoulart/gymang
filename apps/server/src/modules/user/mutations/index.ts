import UserAdd from './UserAddMutation';
import UserLogin from './UserLoginMutation';
import UserLogout from './UserLogoutMutation';

const UserMutations = {
  UserAdd,
  UserLogin,
  UserLogout,
};

export default UserMutations;
