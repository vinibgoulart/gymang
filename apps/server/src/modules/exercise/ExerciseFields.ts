import type { GraphQLContext } from '@gymang/core';
import { ExerciseFilterInputType, ExerciseLoader } from '@gymang/exercise';
import { edgeField, connectionArgs, withFilter } from '@gymang/graphql';
import { GraphQLNonNull } from 'graphql';
import type { ConnectionArguments } from 'graphql-relay';

import ExerciseType, { ExerciseConnection } from './ExerciseType';

export const exerciseTypeField = (
  key = 'exercise',
  bypassViewerCanSee = false,
) => ({
  [key]: {
    type: ExerciseType,
    resolve: async (obj, _, context: GraphQLContext) => {
      return ExerciseLoader.load(context, obj[key], bypassViewerCanSee);
    },
  },
});

export const exerciseEdgeField = () =>
  edgeField({
    connection: ExerciseConnection,
    load: ExerciseLoader.load,
    edgeName: 'exerciseEdge',
    name: 'Exercise',
  });

type MeExercisesConnectionArgs = {
  filters?: ExerciseFilterInputType;
} & ConnectionArguments;

export const exercisesConnectionField = <T = unknown>(
  customResolver = null,
) => ({
  exercises: {
    type: new GraphQLNonNull(ExerciseConnection.connectionType),
    args: {
      ...connectionArgs,
      filters: {
        type: ExerciseFilterInputType,
      },
    },
    resolve: (
      obj: T,
      args: MeExercisesConnectionArgs,
      context: GraphQLContext,
    ) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      const argsWithWorkoutSplit = withFilter(args, {
        workoutSplit: args.filters?.workoutSplit,
      });

      return ExerciseLoader.loadAll(context, argsWithWorkoutSplit);
    },
  },
});
