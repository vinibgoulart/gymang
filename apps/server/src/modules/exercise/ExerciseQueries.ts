import { exercisesConnectionField } from './ExerciseFields';

const ExerciseQueries = {
  ...exercisesConnectionField(),
};

export default ExerciseQueries;
