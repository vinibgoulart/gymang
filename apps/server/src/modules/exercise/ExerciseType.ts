import type { GraphQLContext } from '@gymang/core';
import type { IExercise } from '@gymang/exercise';
import {
  ExerciseLoader,
  ExerciseMuscleGroupEnum,
  getLastSession,
  getSessionInProgress,
} from '@gymang/exercise';
import {
  connectionDefinitions,
  nodeInterface,
  registerTypeLoader,
} from '@gymang/graphql';
import { GraphQLNonNull, GraphQLObjectType, GraphQLString } from 'graphql';
import { globalIdField } from 'graphql-relay';

import { sessionConnectionField } from './session/SessionFields';
import SessionType from './session/SessionType';
import { userTypeField } from '../user/UserFields';
import { workoutSplitTypeField } from '../workoutSplit/WorkoutSplitFields';

const ExerciseType = new GraphQLObjectType<IExercise, GraphQLContext>({
  name: 'Exercise',
  description: 'Represents a Exercise',
  fields: () => ({
    id: globalIdField('Exercise'),
    name: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (exercise) => exercise.name,
    },
    series: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (exercise) => exercise.series,
    },
    repetitions: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (exercise) => exercise.repetitions,
    },
    breakTime: {
      type: GraphQLString,
      resolve: (exercise) => exercise.breakTime,
    },
    muscleGroup: {
      type: new GraphQLNonNull(ExerciseMuscleGroupEnum),
      resolve: (exercise) => exercise.muscleGroup,
    },
    weight: {
      type: GraphQLString,
      resolve: (exercise) => exercise.weight,
    },
    sessionInProgress: {
      type: SessionType,
      resolve: (exercise) => getSessionInProgress({ exercise }),
    },
    lastSession: {
      type: SessionType,
      resolve: (exercise) => getLastSession({ exercise }),
    },
    ...sessionConnectionField(),
    ...userTypeField(),
    ...workoutSplitTypeField(),
  }),
  interfaces: () => [nodeInterface],
});

export const ExerciseConnection = connectionDefinitions({
  name: 'Exercise',
  nodeType: ExerciseType,
});

registerTypeLoader(ExerciseType, ExerciseLoader.load);

export default ExerciseType;
