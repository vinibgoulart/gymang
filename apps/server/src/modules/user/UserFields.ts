import { GraphQLNonNull } from 'graphql';
import UserType, { UserConnection } from './UserType';
import { UserLoader } from '@gymang/user';
import {
  edgeField,
  connectionArgs,
  NullConnection,
  withFilter,
} from '@gymang/graphql';
import { isLoggedIn } from '@gymang/core';

export const userTypeField = (key = 'user', bypassViewerCanSee = false) => ({
  [key]: {
    type: UserType,
    resolve: async (obj, args, context) => {
      return UserLoader.load(context, obj[key], bypassViewerCanSee);
    },
  },
});

export const userEdgeField = () =>
  edgeField({
    connection: UserConnection,
    load: UserLoader.load,
    edgeName: 'userEdge',
    name: 'User',
  });

export const userConnectionField = (customResolver = null) => ({
  users: {
    type: new GraphQLNonNull(UserConnection.connectionType),
    args: {
      ...connectionArgs,
      // filters: {
      //   type: UserFilterInputType,
      // },
    },
    resolve: (obj, args, context) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      if (!isLoggedIn(context)) {
        return NullConnection;
      }

      const argsWithCompany = withFilter(args, {
        merchant: context?.merchant?._id,
      });

      return UserLoader.loadAll(context, argsWithCompany);
    },
  },
});

export const meField = () => ({
  me: {
    type: UserType,
    resolve: (_root, _args, context) => {
      return UserLoader.load(context, context.user?._id);
    },
  },
});
