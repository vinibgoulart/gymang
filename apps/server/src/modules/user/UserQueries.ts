import { meField, userConnectionField } from './UserFields';

const UserQueries = {
  ...meField(),
  ...userConnectionField(),
};

export default UserQueries;
