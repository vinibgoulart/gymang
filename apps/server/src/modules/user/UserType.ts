import { GraphQLContext } from '@gymang/core';
import {
  connectionDefinitions,
  nodeInterface,
  registerTypeLoader,
} from '@gymang/graphql';
import { IUser, UserLoader } from '@gymang/user';
import { GraphQLObjectType, GraphQLString } from 'graphql';
import { globalIdField } from 'graphql-relay';

const UserType = new GraphQLObjectType<IUser, GraphQLContext>({
  name: 'User',
  description: 'Represents a User',
  fields: () => ({
    id: globalIdField('User'),
    firstName: {
      type: GraphQLString,
      resolve: (user) => user.firstName,
    },
    email: {
      type: GraphQLString,
      resolve: (user) => user.email,
    },
  }),
  interfaces: () => [nodeInterface],
});

export const UserConnection = connectionDefinitions({
  name: 'User',
  nodeType: UserType,
});

registerTypeLoader(UserType, UserLoader.load);

export default UserType;
