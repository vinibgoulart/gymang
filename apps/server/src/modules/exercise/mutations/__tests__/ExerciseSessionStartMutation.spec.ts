import { GRAPHQL_TYPE } from '@gymang/core';
import { Exercise, handleCreateExercise } from '@gymang/exercise';
import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  sanitizeTestObject,
} from '@gymang/testutils';
import { handleCreateUser } from '@gymang/user';
import { handleCreateWorkoutSplit } from '@gymang/workout-split';
import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import { getContext } from '../../../../getContext';
import { schema as schemaWeb } from '../../../../graphql/schema/schema';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

const query = `
  mutation ExerciseSessionStartSpecMutation($input: ExerciseSessionStartInput!) {
    ExerciseSessionStart(input: $input) {
      exercise {
        id
        name
        user {
          firstName
        }
        workoutSplit {
          id
          name
        }
        series
        repetitions
        breakTime
        muscleGroup
        weight
      }
      success
      error
    }
  }
`;

it('should start a exercise session', async () => {
  const user = await handleCreateUser();
  const exercise = await handleCreateExercise();

  const input = {
    id: toGlobalId('Exercise', exercise._id),
  };

  const variables = {
    input,
  };

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaWeb,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.ExerciseSessionStart.error).toBeNull();
  expect(result.data.ExerciseSessionStart.success).toEqual(
    'Session created successfully',
  );

  const exerciseUpdated = await Exercise.findOne({
    _id: exercise._id,
  });

  expect(exerciseUpdated.sessions).toHaveLength(1);

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});

it('should not start a session for a exercise from another user', async () => {
  const user = await handleCreateUser();
  const anotherUser = await handleCreateUser();

  const exercise = await handleCreateExercise({
    user: anotherUser,
  });

  const input = {
    id: toGlobalId('Exercise', exercise._id),
  };

  const variables = {
    input,
  };

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaWeb,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.ExerciseSessionStart.error).toEqual(
    'You can not initiate a Session for this exercise',
  );
  expect(result.data.ExerciseSessionStart.success).toBeNull();

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});

it('should not start a session for a exercise with a in progress session', async () => {
  const user = await handleCreateUser();

  const exercise = await handleCreateExercise({
    sessions: [
      {
        series: '1',
        repetitions: '1',
        breakTime: '1',
        muscleGroup: 'CHEST',
        weight: '1',
        finishedAt: null,
      },
    ],
  });

  const input = {
    id: toGlobalId('Exercise', exercise._id),
  };

  const variables = {
    input,
  };

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaWeb,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.ExerciseSessionStart.error).toEqual(
    'You already have a session in progress, finish it first',
  );
  expect(result.data.ExerciseSessionStart.success).toBeNull();

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});

it('should start a session for a exercise with a completed session', async () => {
  const user = await handleCreateUser();

  const exercise = await handleCreateExercise({
    sessions: [
      {
        series: '1',
        repetitions: '1',
        breakTime: '1',
        muscleGroup: 'CHEST',
        weight: '1',
        finishedAt: new Date(),
      },
    ],
  });

  const input = {
    id: toGlobalId('Exercise', exercise._id),
  };

  const variables = {
    input,
  };

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaWeb,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.ExerciseSessionStart.error).toBeNull();
  expect(result.data.ExerciseSessionStart.success).toEqual(
    'Session created successfully',
  );

  const exerciseUpdated = await Exercise.findOne({
    _id: exercise._id,
  });

  expect(exerciseUpdated.sessions).toHaveLength(2);

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});

it('should not start a session for a workout split with an exercise in progress', async () => {
  const user = await handleCreateUser();
  const workoutSplit = await handleCreateWorkoutSplit();

  await handleCreateExercise({
    workoutSplit,
    sessions: [
      {
        series: '1',
        repetitions: '1',
        breakTime: '1',
        muscleGroup: 'CHEST',
        weight: '1',
        finishedAt: null,
      },
    ],
  });

  const exercise = await handleCreateExercise({
    workoutSplit,
  });

  const input = {
    id: toGlobalId('Exercise', exercise._id),
  };

  const variables = {
    input,
  };

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaWeb,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.ExerciseSessionStart.error).toEqual(
    'You already have a session in progress, finish it first',
  );
  expect(result.data.ExerciseSessionStart.success).toBeNull();

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});
