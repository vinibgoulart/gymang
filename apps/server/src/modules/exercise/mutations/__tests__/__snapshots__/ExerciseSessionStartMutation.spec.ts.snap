// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not start a session for a exercise from another user 1`] = `
{
  "ExerciseSessionStart": {
    "error": "You can not initiate a Session for this exercise",
    "exercise": "EMPTY",
    "success": "EMPTY",
  },
}
`;

exports[`should not start a session for a exercise with a in progress session 1`] = `
{
  "ExerciseSessionStart": {
    "error": "You already have a session in progress, finish it first",
    "exercise": "EMPTY",
    "success": "EMPTY",
  },
}
`;

exports[`should not start a session for a workout split with an exercise in progress 1`] = `
{
  "ExerciseSessionStart": {
    "error": "You already have a session in progress, finish it first",
    "exercise": "EMPTY",
    "success": "EMPTY",
  },
}
`;

exports[`should start a exercise session 1`] = `
{
  "ExerciseSessionStart": {
    "error": "EMPTY",
    "exercise": {
      "breakTime": "EMPTY",
      "id": "FROZEN-ID",
      "muscleGroup": "BACK",
      "name": "Exercise NaN name",
      "repetitions": "12",
      "series": "4",
      "user": {
        "firstName": "User 1 firstName",
      },
      "weight": "EMPTY",
      "workoutSplit": {
        "id": "FROZEN-ID",
        "name": "WorkoutSplit NaN name",
      },
    },
    "success": "Session created successfully",
  },
}
`;

exports[`should start a session for a exercise with a completed session 1`] = `
{
  "ExerciseSessionStart": {
    "error": "EMPTY",
    "exercise": {
      "breakTime": "EMPTY",
      "id": "FROZEN-ID",
      "muscleGroup": "BACK",
      "name": "Exercise 3 name",
      "repetitions": "12",
      "series": "4",
      "user": {
        "firstName": "User 1 firstName",
      },
      "weight": "EMPTY",
      "workoutSplit": {
        "id": "FROZEN-ID",
        "name": "WorkoutSplit 1 name",
      },
    },
    "success": "Session created successfully",
  },
}
`;
